alea/runner.py

```diff
@@ -0,0 +1,400 @@
+        parameter_list = sorted(self.statistical_model.get_parameter_list())
+        result_list = parameter_list + ['ll', 'dl', 'ul']
+        result_dtype = [(n, float) for n in parameter_list]
+        result_dtype += [(n, float) for n in ['ll', 'dl', 'ul']]
+        # try:
+        #     parameter_list += self.statistical_model.additional_parameters
+        #     result_dtype += [(n, float) for n in self.statistical_model.additional_parameters]
+        # except:
+        #     pass
+        return parameter_list, result_list, result_dtype
+
+    def _get_generate_values(self):
+        """Get generate values list from hypotheses"""
+        generate_values = []
+        hypotheses = deepcopy(self.hypotheses)
+
+        for hypothesis in hypotheses:
+            if hypothesis == 'null':
+                # there is no signal component
+                hypothesis = {self.poi: 0.}
+            elif hypothesis == 'true':
+                # the true signal component is used
+                if self.poi not in self.true_generate_values:
+                    raise ValueError(
+                        f'{self.poi} should be provided in true_generate_values',
+                    )
+                hypothesis = {
+                    self.poi: self.true_generate_values.get(self.poi),
+                }
+            elif hypothesis == 'free':
+                hypothesis = {}
+
+            array = deepcopy(self.common_generate_values)
+            array.update(hypothesis)
+            generate_values.append(array)
+        return generate_values
+
+    def write_output(self, results):
+        """Write output file with metadata"""
+        metadata = deepcopy(self.metadata)
+
+        result_names = [f'{i:d}' for i in range(len(self.generate_values))]
+        for i, ea in enumerate(self.hypotheses):
+            if ea in ['null', 'free', 'true']:
+                result_names[i] = ea
+
+        metadata['date'] = datetime.now().strftime('%Y%m%d_%H:%M:%S')
+        metadata['poi'] = self.poi
+        metadata['common_generate_values'] = self.common_generate_values
+        metadata['true_generate_values'] = self.true_generate_values
+
+        array_metadatas = [{'generate_values': ea} for ea in self.generate_values]
+
+        numpy_arrays_and_names = [(r, rn) for r, rn in zip(results, result_names)]
+
+        print(f'Saving {self.output_file}')
+        numpy_to_toyfile(
+            self.output_file,
+            numpy_arrays_and_names=numpy_arrays_and_names,
+            metadata=metadata,
+            array_metadatas=array_metadatas)
+
+    def read_toydata(self):
+        """Read toydata from file"""
+        toydata, toydata_names = toydata_from_file(self.toydata_file)
+        return toydata, toydata_names
+
+    def toy_simulation(self):
+        """
+        Run toy simulation a specified different toydata mode
+        and loop over generate values
+        """
+        flag_read_toydata = False
```

Comment:
I think it is more readable if we do not use these bool flags, but just check toydata_mode each time?

Reply:
updated in previous commits.